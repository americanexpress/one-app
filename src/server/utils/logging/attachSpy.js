/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export default function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function monkeypatchedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}
