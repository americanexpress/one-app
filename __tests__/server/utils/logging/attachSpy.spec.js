/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import attachSpy from '../../../../src/server/utils/logging/attachSpy';

describe('attachSpy', () => {
  it('throws if the method name is not a function on the object', () => {
    expect(() => attachSpy({}, 'method', () => {})).toThrowErrorMatchingSnapshot();
  });
  it('throws if the spy is not a function', () => {
    expect(() => attachSpy({ method: () => {} }, 'method', 'hello')).toThrowErrorMatchingSnapshot();
  });
  describe('monkeypatched method', () => {
    const originalMethod = jest.fn(() => 'some return value');
    const obj = {};

    beforeEach(() => {
      obj.method = originalMethod;
      originalMethod.mockClear();
    });

    it('invokes the spy', () => {
      const spy = jest.fn();
      attachSpy(obj, 'method', spy);
      expect(spy).not.toHaveBeenCalled();
      obj.method();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('invokes the spy with args', () => {
      const spy = jest.fn();
      attachSpy(obj, 'method', spy);
      obj.method('g', { h: 'i' }, 9);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy.mock.calls[0][0]).toEqual(['g', { h: 'i' }, 9]);
    });

    it('invokes the spy with callOriginal', () => {
      const spy = jest.fn();
      attachSpy(obj, 'method', spy);
      obj.method();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(typeof spy.mock.calls[0][1]).toEqual('function');
    });

    it('returns the original methods return value from callOriginal', () => {
      expect.assertions(1);
      attachSpy(obj, 'method', (args, callOriginal) => {
        expect(callOriginal()).toEqual('some return value');
      });
      obj.method();
    });

    it('calls the original method when the spy does not', () => {
      attachSpy(obj, 'method', () => {});
      obj.method();
      expect(originalMethod).toHaveBeenCalledTimes(1);
    });

    it('does not call the original method when the spy already had', () => {
      attachSpy(obj, 'method', (args, callOriginal) => {
        callOriginal();
      });
      obj.method();
      expect(originalMethod).toHaveBeenCalledTimes(1);
    });

    it('returns the original methods return value to the caller of the monkeypatched method', () => {
      attachSpy(obj, 'method', jest.fn());
      expect(obj.method()).toEqual('some return value');
    });
  });
});
