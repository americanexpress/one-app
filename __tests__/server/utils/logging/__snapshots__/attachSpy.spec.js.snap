// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachSpy throws if the method name is not a function on the object 1`] = `"method is not a function"`;

exports[`attachSpy throws if the spy is not a function 1`] = `"spy must be a function (was "string")"`;
